File: linked_list/206_reverse_linked_list.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def reverse_list(head: Optional[ListNode]) -> Optional[ListNode]:
    if not head:
        return head

    current = head
    prev = None
    while current:
        old_next = current.next
        current.next = prev
        prev = current
        current = old_next

    return prev

File: linked_list/206_reverse_linked_list.go
type ListNode struct {
	Val  int
	Next *ListNode
}

func reverseList(head *ListNode) *ListNode {
	// если сделать тут просто prev := &ListNode{}
	var prev *ListNode
	current := head
	for current != nil {
		tmpNext := current.Next
		current.Next = prev
		prev = current
		current = tmpNext
	}

	return prev
}